### PR TITLE
multi: Use atomic types in unexported modules.

### PR DESCRIPTION
**Pull request by jholdstock**
go 1.19 introduced [new atomic types](https://go.dev/doc/go1.19#atomic_types) which hide the underlying primitive types so that all accesses are forced to use the atomic APIs. This makes the code less prone to human error and a bit less verbose.

Of particular note is the introduction of the type `atomic.Bool`. Previously this repo made use of atomic integers to track properties with boolean values, but this PR updates those to use `atomic.Bool` instead.

### CODE REVIEW

File: peer/peer.go [hunk 1]
```diff
@@ -1967,7 +1964,7 @@ func (p *Peer) start() error {
 // have no effect.
 func (p *Peer) AssociateConnection(conn net.Conn) {
 	// Already connected?
-	if !atomic.CompareAndSwapInt32(&p.connected, 0, 1) {
+	if !p.connected.CompareAndSwap(false, true) {
```
Comment (jholdstock) [inline peer/peer.go:1967]: I like how `CompareAndSwap` was being used here to ensure actions are only performed once. I have copied this pattern to other places where it is appropriate (once in this file, twice in `addrmanager.go` and once in `rpcwebsocket.go`)

File: internal/mining/cpuminer/cpuminer.go [hunk 2]
```diff
@@ -826,14 +826,15 @@ out:
...
 		g:                 cfg.BgBlkTmplGenerator,
 		cfg:               cfg,
-		numWorkers:        defaultNumWorkers,
 		updateNumWorkers:  make(chan struct{}),
 		queryHashesPerSec: make(chan float64),
 		speedStats:        make(map[uint64]*speedStats),
 		minedOnParents:    make(map[chainhash.Hash]uint8),
 		quit:              make(chan struct{}),
 	}
+	miner.numWorkers.Store(defaultNumWorkers)
```
Comment (jholdstock) [inline internal/mining/cpuminer/cpuminer.go:838]: This change is probably worth extra scrutiny.

Previously the underlying `numWorkers` primitive was set directly **during** creation of the `CPUMiner`, but that is no longer possible so it is instead set using the proper atomic `Store` func **after** creation.

File: connmgr/connmanager.go [hunk 3]
```diff
@@ -307,12 +302,12 @@ out:
 			case registerPending:
 				connReq := msg.c
 				connReq.updateState(ConnPending)
-				pending[msg.c.id] = connReq
+				pending[msg.c.id.Load()] = connReq
```
Comment (jholdstock) [inline connmgr/connmanager.go:305]: Previously the loop inside the `connHandler` function was accessing the `ConnReq.id` primitive directly rather than via atomic funcs.

File: connmgr/connmanager.go [hunk 4]
```diff
@@ -530,8 +525,8 @@ func (cm *ConnManager) Connect(ctx context.Context, c *ConnReq) {
 		return
 	}
 
-	if atomic.LoadUint64(&c.id) == 0 {
-		atomic.StoreUint64(&c.id, atomic.AddUint64(&cm.connReqCount, 1))
+	if c.id.Load() == 0 {
+		c.id.Store(cm.connReqCount.Add(1))
```
Comment (jrick) [inline connmgr/connmanager.go:529]: not the fault of your patch, but this looks very racy.  c.id may not longer be 0 by the time we store the value.

Comment (jrick) [reply connmgr/connmanager.go:529]: I don't see a straightforward way to implement this correctly using atomics.  Both c.id and cm.connReqCount need to be updated atomically together.  This might require a mutex instead.

File: internal/mining/cpuminer/cpuminer.go (hunk 2 above)
Comment (jrick) [reply internal/mining/cpuminer/cpuminer.go:838]: this is fine. the pointer has not been returned to anyone else yet.

**Review by davecgh — CHANGES_REQUESTED** [review]
Thanks for the PR.

Note that this change would have to bump the relevant `go.mod` files to require Go 1.19 because the types were not introduced until Go 1.19.

I don't have any objections to the internal instances since they only require the bump to the main `dcrd` module, but I'd prefer not to needlessly require newer Go versions for the other modules which hoist everyone else forward unless there is a really good reason and I don't think this qualifies.

So, in short, no objections for all of the one that touch the internal packages and code in the main module. However, I would prefer not to hoist everyone forward (at least yet) for the others.

**Comment by jrick** [comment]
Go <=1.18 no longer receives security patches.  I'm fine leaving it behind.

File: internal/mining/cpuminer/cpuminer.go (hunk 2 above)
Comment (davecgh) [reply internal/mining/cpuminer/cpuminer.go:838]: Agreed.  There is no issue here as the pointer is not accessible to the caller yet as it hasn't been returned.

**Review by davecgh — COMMENTED** [review]
I reviewed the changes to `internal/*` and `server.go` and they look good.

I didn't review the ones to `addrmgr`, `connmgr`, `database`, `peer`, or `rpcclient` since they're in separate modules, and as previously mentioned, I don't want to hoist everyone forward for those.

**Comment by davecgh** [comment]
> Go <=1.18 no longer receives security patches. I'm fine leaving it behind.

That's a choice each consumer has to make with their own projects, imo.  I don't think it's fair to force our own policies on everyone who uses our modules just to get updated versions of them in general.  As someone who has been bitten many times by being on the receiving end of such policies, I'm particularly sensitive to it.

For any cases where there is a legitimate security issue resolved that genuinely requires a new version, we'd obviously bump them.  

It might also be reasonable to follow the Go bootstrap toolchain as a minimum here since you won't even be able to build the relevant version of Go without that version.

**Review by davecgh — APPROVED** [review]

**Comment by jholdstock** [comment]
Exported modules are now updated in #3054.

I've rolled the changes in this PR into one commit because they are all under a single `go.mod`.

File: connmgr/connmanager.go (hunk 4 above)
Comment (jholdstock) [reply connmgr/connmanager.go]: Moved to #3054
